### DISCUSSION TITLE
Extract Event Args Utility Type

### DISCUSSION THREAD

**Discussion by fdarian** [wagmi-dev/abitype#69] [Ideas]
Hi,

I have an idea on `ExtractEventArgs` utility types that might solves common problem on typing event.

## Motivation

Coming from wagmi and TypeChain, as far as I know there isn't any way to extract fully typed event(s) from logs inside transaction receipt, assuming the contract/ABI has already been known. There are actually two possible solutions for this,

1. Using TypeChain generated event, here is an example of parsing receipt's log:

<img width="640" alt="Screenshot 2023-01-04 at 16 54 11" src="https://user-images.githubusercontent.com/62016900/210529163-1f83e533-a9e8-4c4b-b0f7-504f6b8525de.png">

2. Using wagmi's [watchContractEvent](https://wagmi.sh/core/actions/watchContractEvent#usage)

The first solution might do the job, but we still need to hardcode the event name and manually cast to the desired event. Meanwhile the second solution which uses `AbiType` under the hood only facilitates "watching events" use-case, I still haven't found the AbiType's solution to type event from logs. 

## Solution

Therefore, I'm experimenting on `ExtractEventArgs` utility type that is used to extract [`AbiEvent`](https://github.com/wagmi-dev/abitype#abievent)'s arguments by converting `AbiEvent.inputs` to  a typed object. It also works alongside the [`ExtractAbiEvent`](https://github.com/wagmi-dev/abitype#extractabievent). 

#### Here's an example that parses ERC20 events.

https://user-images.githubusercontent.com/62016900/210531555-c830f3a1-19a9-48bd-9ce6-2e06390c748f.mov

#### Here's an example to type ethers [LogDescription](https://docs.ethers.org/v5/api/utils/abi/interface/#LogDescription).

https://user-images.githubusercontent.com/62016900/210533804-8d639bb8-fd63-4422-8e51-c0ea83595613.mov

From the example above, I still need to cast to `any` first since I just quickly use TypeChain's generated code. In future, this helper type could be integrated to any upcoming wagmi features that provide automatic type inference. Something like `getContractInferface(<abi>)` or `paseLogs(<abi>)`.

It's also possible to improve the current [`event`](https://wagmi.sh/react/hooks/useContractEvent#listener) or [`args`](https://wagmi.sh/core/actions/readContract#args-optional) that's currently still uses array. However, this might have some reasons that I'm not aware of why wagmi chose to use array instead of object.

Let me know what you guys think, happy to add a PR if this is useful.

## Edited

Here's a better example on parsing log

![Screenshot 2023-01-04 at 18 55 36](https://user-images.githubusercontent.com/62016900/210549944-f51c1094-7887-4d5d-93f0-689dcc06417a.png)

**Comment by fdarian** [comment]
> Hi,
> 
> I have an idea on `ExtractEventArgs` utility types that might solves common problem on typing event.
> 
> ## Motivation
> Coming from wagmi and TypeChain, as far as I know there isn't any way to extract fully typed event(s) from logs inside transaction receipt, assuming the contract/ABI has already been known. There are actually two possible solutions for this,
> 
> 1. Using TypeChain generated event, here is an example of parsing receipt's log:
> 
> <img alt="Screenshot 2023-01-04 at 16 54 11" width="640" src="https://user-images.githubusercontent.com/62016900/210529163-1f83e533-a9e8-4c4b-b0f7-504f6b8525de.png">
> 
> 2. Using wagmi's [watchContractEvent](https://wagmi.sh/core/actions/watchContractEvent#usage)
> 
> The first solution might do the job, but we still need to hardcode the event name and manually cast to the desired event. Meanwhile the second solution which uses `AbiType` under the hood only facilitates "watching events" use-case, I still haven't found the AbiType's solution to type event from logs.
> 
> ## Solution
> Therefore, I'm experimenting on `ExtractEventArgs` utility type that is used to extract [`AbiEvent`](https://github.com/wagmi-dev/abitype#abievent)'s arguments by converting `AbiEvent.inputs` to a typed object. It also works alongside the [`ExtractAbiEvent`](https://github.com/wagmi-dev/abitype#extractabievent).
> 
> #### Here's an example that parses ERC20 events.
>  Screen.Recording.2023-01-04.at.17.06.42.mov 
> #### Here's an example to type ethers [LogDescription](https://docs.ethers.org/v5/api/utils/abi/interface/#LogDescription).
>  Screen.Recording.2023-01-04.at.17.18.12.mov 
> From the example above, I still need to cast to `any` first since I just quickly use TypeChain's generated code. In future, this helper type could be integrated to any upcoming wagmi features that provide automatic type inference. Something like `getContractInferface(<abi>)` or `paseLogs(<abi>)`.
> 
> It's also possible to improve the current [`event`](https://wagmi.sh/react/hooks/useContractEvent#listener) or [`args`](https://wagmi.sh/core/actions/readContract#args-optional) that's currently still uses array. However, this might have some reasons that I'm not aware of why wagmi chose to use array instead of object.
> 
> Let me know what you guys think, happy to add a PR if this is useful.

I just remember with TypeChain we can also use `.queryFilter`. But it requires `contract.filter.<event-name>()`, not handy if we need to parse transaction receipt coming from `.wait()`for example.

**Comment by gsteenkamp89** [comment]
@farreldarian 
This is really awesome! 
Not sure if you ever made this PR but I'd certainly make use of this if you're happy to share your code! 🙏

**Reply by fdarian** [reply]
Glad that you found it helpful!

I actually still waiting from @jxom, @tmm, or the rest from the community, whether this could be just an util, or maybe a breaking change for the next version.

**Reply by fdarian** [reply]
Quick correction, I forgot that Viem already [has this feature already.](https://github.com/wagmi-dev/abitype/discussions/69#discussioncomment-7490548)

Anyway, where did you find it to be usefull? (assuming you are using viem)

**Comment by fdarian** [comment]
### UPDATE 10 months later

![CleanShot 2023-11-07 at 00 48 55@2x](https://github.com/wagmi-dev/abitype/assets/62016900/81eea97f-f558-4b8f-91dd-31b7e0900a7d)

10 months back we didn't have viem, but now the [`getContractEvents().args`](https://viem.sh/docs/contract/getContractEvents.html#args) is fully typed!

I think this discussion can be closed

**Comment by gsteenkamp89** [comment]
I still needed this type helper so I did eventually find it. For those interested, Viem exports `GetEventArgs`.
One thing to note is that by default it excludes non-indexed args, so to get the exact same type that you get when calling useContractEvent:

```typescript
import { GetEventArgs } from "viem";

type EventArgs = GetEventArgs<
  typeof abi,
  "EventName",
  { IndexedOnly: false } 
>;
```
